rfcs/003 - Multi-Region Support RFC.md

```diff
@@ -0,0 +1,161 @@
+# 003 - Multi-Region Suppor RFC
+
+## Current Status
+
+### Proposed
+
+2022-08-22
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Luis Fernando de Pombo <l@iasql.com>
+- Alejandro Guillen <alejandro@iasql.com>
+- Yolanda Robla <yolanda@iasql.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- David Ellis <david@iasql.com>
+
+## Summary
+
+Due to the structure of the AWS API, as well as some hesitance to accidentally over-complicate our tables, we implemented the root `aws_account` module to require a specific AWS Region to be chosen for the database. Thus to manage your entire AWS account you would need a separate database for every region you're using, and be careful not to install modules for services that are "global" in scope like Route53 on more than one of these databases. It is also impossible to safely represent cross-region entities like DynamoDB clusters. This is an untenable situation for IaSQL for serious usage, and so enabling management across all regions within a single database must be tackled.
+
+## Proposal
+
+After going through several options in many directions, we've chosen adding a `region` column to all relevant tables, but with a `default` value provided in the schema so users can normally ignore that column if they deploy primarily to just one region.
+
+In order to keep the current flexibility, there would be a Postgres function provided by the `aws_account` module to return the default region they have selected. This could simply be a rename of the current `region` column on the `aws_account` table to `default_region`, but we also want to improve on the current state of things when moving to multi-region modules.
+
+The first thing that we desire is that the default region be guaranteed to be an actual region that the user is able to access. That depends on their credentials, so we should have an `aws_regions` table contains the list of all regions that we acquire via the API, and then joined in on the `default_region` column.
+
+The second thing that is necessary with this approach is that the default region function output needs to always return a region.
+
+These two are not compatible with each other, and here's why:
+
+If we put a foreign key join on the `aws_account` table to the `aws_regions` table, then when we first attempt to insert the credentials to make it available to the Mapper in charge of the `aws_regions` table, we would have to have the `default_region` be set to `NULL` to not fall afoul of the foreign key constraints. But if it is `NULL` then the default value function would need to have some other value it always spits out to users which we don't *know* for sure could be something like `us-east-1` because it depends on the response for the regions available to the credentials in question.
```

Comment:
I take it from this that AWS regions available vary per account? Perhaps we can bootstrap the initial list with default list of readily accessible regions from AWS and once we get credentials we can update it to get over this hump?

Reply:
Read a bit further on, I have a solution already :)